### DISCUSSION TITLE
Pico Servo Travel

### DISCUSSION THREAD

**Discussion by iansmiler** [micropython/micropython#13878] [RP2040 / Pico]
Hello, 

I am learning to use micropython with servos. When I initialise my code, the servo jumps to 0 degrees at maximum velocity (I'm guessing it's zero degrees), before executing the routine. How can I make it travel from the position it finds itself in to the 90-degree position without going to zero first? and ideally at the speed I set which is dictated by delay_ms?

Any help is appreciated. here is my code:

```
import machine
import time
from servo import Servo

# Create our Servo object, assigning the
# GPIO pin connected the PWM wire of the servo
my_servo = Servo(pin_id=16)
delay_ms = 12  # Amount of milliseconds to wait between servo movements
servoCpos = 90
        
def positioncentre(delay_ms):
    current_position = my_servo.read()
    target_position = servoCpos  # Center position
    
    # Gradually move the servo towards the center position
    while current_position != target_position:
        if current_position < target_position:
            current_position += 1
        else:
            current_position -= 1
        
        my_servo.write(current_position)
        time.sleep_ms(delay_ms)
        
positioncentre(delay_ms)

```

**Comment by scruss** [comment]
please edit/repost, but [Quoting code](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) properly. We can't make out what you're trying to do

**Reply by iansmiler** [reply]
Okay - I just edited it - have I done it correctly this time?

**Comment by rkompass** [comment]
Without having read the library code I suppose:
 that `my_servo.read()` simply reads the internal position variable of the Servo class, which is 0 after booting/initialization.
So this `my_servo.read()` is not a true reading of the physical position.

Analog servo positions cannot be read. The servo is following a PWM-like signal, which is only a one-way communication.
As this is done automatically with timers, like in PWM, after initialization the servo goes to position 0, as this is the position it has internally stored at first (in e.g. a variable `self.position`).

You might try to init the servo (or change the Servo classes init to allow it to) so that you specify a position different from 0 at initialization. Assuming that you know the physical position of the servo at the start.

**Reply by iansmiler** [reply]
Thank you for the education! I'm learning a lot with this one 👍 I am happy for my servo to move to the point of reference at first initiation, but it is going to be used with a sonar module HC-SR04 to:

look left (Take distance reading and store)
Look right (Take distance reading and store)
move back to the centre position (Take distance reading and store)

I will then use the stored values elsewhere in the program.

This cycle will repeat every 5 seconds. I will have a play with the program now to try and combat the fast initiation movement of called sub-routines.

Interestingly - now that I have added the code for the HC-SR04, the positioncheck sub-routine seems to be not using the delay_ms variable as a reference - the routine is moving much slower than the positioncheck routine. 

Code - thus far:

```
import machine
import time
from servo import Servo

# Create our Servo object, assigning the
# GPIO pin connected the PWM wire of the servo
my_servo = Servo(pin_id=16)
delay_ms = 6  # Amount of milliseconds to wait between servo movements
servoLpos = 34
servoRpos = 156
servoCpos = 94

# HC-SR04 pins
trigger_pin = machine.Pin(14, machine.Pin.OUT)
echo_pin = machine.Pin(15, machine.Pin.IN)

# List to store distance measurements at each servo position
distance_measurements = []

def measure_distance():
    # Send a 10 microsecond pulse to trigger_pin
    trigger_pin.value(1)
    time.sleep_us(10)
    trigger_pin.value(0)
    
    # Measure the duration of the echo pulse
    duration = machine.time_pulse_us(echo_pin, 1, 30000)  # Timeout set to 30ms (max range of HC-SR04)
    
    # Convert the duration to distance (in centimeters)
    distance = duration / 58  # Conversion formula for HC-SR04
    
    return distance

def positioncheck():
    # Temporary list to store distance measurements for this movement
    movement_measurements = []
    
    for position in range(servoLpos, servoRpos):  # Step the position forward from 0deg to 180deg
        my_servo.write(position)  # Set the Servo to the current position
        time.sleep_ms(delay_ms)  # Wait for the servo to make the movement
        distance = measure_distance()
        movement_measurements.append(distance)  # Store distance measurement
        
    
    # Add the measurements of this movement to the main list
    distance_measurements.append(movement_measurements)

def positioncentre(delay_ms):
    current_position = my_servo.read()
    target_position = servoCpos  # Center position
    
    # Gradually move the servo towards the center position
    while current_position != target_position:
        if current_position < target_position:
            current_position += 1
        else:
            current_position -= 1
        
        my_servo.write(current_position)
        time.sleep_ms(delay_ms)
    
    # Measure distance at center position and store it directly in the main list
    distance = measure_distance()
    distance_measurements.append(distance)

# Example usage
positioncheck()
positioncentre(delay_ms)

# Print the stored distance measurements
print("Stored distance measurements:")
for movement_measurements in distance_measurements:
    print(movement_measurements)
```
